File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        counter = Counter(word).most_common()
        numDict = defaultdict(list)
        result = 0
        idx = 0

        for letter, cnt in counter:
            numDict[idx].append(letter)
            result += len(numDict[idx]) * cnt
            idx = (idx + 1) % 8

        return result

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.java
class Solution {
    public int minimumPushes(String word) {
        int[] letters = new int[26];

        for (int i = 0; i < word.length(); i++) {
            letters[word.charAt(i) - 'a'] += 1;
        }

        Arrays.sort(letters);

        int idx = 1, cnt = 1, result = 0;

        for (int i = 25; i >= 0; i--) {
            if (letters[i] == 0) break;
            if (idx < 9) {
                idx++;
            } else {
                idx = 2;
                cnt++;
            }
            result += letters[i] * cnt;
        }

        return result;
    }
}